File: 127.单词接龙.py
#
# @lc app=leetcode.cn id=127 lang=python3
#
# [127] 单词接龙
#

# @lc code=start
import string
from typing import List


# 14课高级搜索
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0

        begin_set = {beginWord}
        end_set = {endWord}
        dist = 1
        wordList = set(wordList)
        word_len = len(beginWord)

        # BFS starts here
        while begin_set:
            dist += 1
            next_front = set()
            for word in begin_set:
                for i in range(word_len):
                    for c in string.ascii_lowercase:
                        if c != word[i]:
                            new_word = word[:i] + c + word[i+1:]
                            if new_word in end_set:
                                return dist
                            if new_word in wordList:
                                next_front.add(new_word)
                                wordList.remove(new_word)
            begin_set = next_front
            if len(end_set) < len(begin_set):
                begin_set, end_set = end_set, begin_set
        return 0


s = Solution()
s.ladderLength("hit", "cog", ["hot", "dot", "dog", "lot", "log", "cog"])
# @lc code=end

File: 127.单词接龙.ts
/*
 * @lc app=leetcode.cn id=127 lang=typescript
 *
 * [127] 单词接龙
 */

// @lc code=start
function ladderLength(beginWord: string, endWord: string, wordList: string[]): number {
  if (!wordList.includes(endWord)) {
    return 0
  }

  let beginSet = new Set([beginWord])
  let endSet = new Set([endWord])
  let wordListSet = new Set(wordList)
  let dist = 1
  while (beginSet.size) {
    dist += 1
    let nextSet = new Set<string>()
    for (const word of beginSet) {
      for (let i = 0; i < beginWord.length; i++) {
        for (let j = 0; j < 26; j++) {
          let char = String.fromCharCode(j + 97)

          if (char != word[i]) {
            let newWord = word.slice(0, i) + char + word.slice(i + 1)
            if (endSet.has(newWord)) {
              return dist
            }
            if (wordListSet.has(newWord)) {
              nextSet.add(newWord)
              wordListSet.delete(newWord)
            }

          }
        }

      }
    }
    beginSet = nextSet
    if (beginSet.size > endSet.size) {
      [beginSet, endSet] = [endSet, beginSet]
    }
  }
  return 0
};
// console.log('r ->', ladderLength("hit", "cog", ["hot", "dot", "dog", "lot", "log", "cog"]))
// console.log('r ->', ladderLength("hot", "dog", ["hot", "dog"]))
// @lc code=end

